File: Packages/WhoopStore/Sources/WhoopStore/WhoopLiveCapabilities.swift
import Foundation

/// Metrics NOOP can honestly derive from a live WHOOP connection without a cloud/CSV import.
/// Calibrated SpO2 percentage is intentionally absent: the decoded WHOOP optical values are raw ADC,
/// and converting those to a percentage would require a validated calibration curve we do not have.
/// Steps are also absent: the reverse-engineered motion counter is not a validated pedometer.
public enum WhoopLiveCapabilities {
    public static let base: Set<Metric> = [.hr, .hrv, .skinTemp, .sleep, .strainLoad]

    public static func metrics(forModel model: String) -> Set<Metric> {
        _ = model
        return base
    }

    public static func encoded(forModel model: String) -> String {
        metrics(forModel: model).map(\.rawValue).sorted().joined(separator: ",")
    }

    public static func withoutUnvalidatedLiveMetrics(
        _ capabilities: Set<Metric>
    ) -> Set<Metric> {
        capabilities.subtracting([.spo2, .steps])
    }

    public static func stripUnvalidatedLiveTokens(
        fromEncoded encoded: String
    ) -> String {
        let rejected = Set([Metric.spo2.rawValue, Metric.steps.rawValue])
        return encoded.split(separator: ",", omittingEmptySubsequences: true)
            .map(String.init)
            .filter { !rejected.contains($0) }
            .joined(separator: ",")
    }

    /// Retained for the historical migration that removed only the SpO2 token.
    public static func withoutCalibratedSpo2(_ capabilities: Set<Metric>) -> Set<Metric> {
        capabilities.subtracting([.spo2])
    }

    /// Retained for the historical migration that removed only the SpO2 token.
    public static func stripSpo2Token(fromEncoded encoded: String) -> String {
        encoded.split(separator: ",", omittingEmptySubsequences: true)
            .map(String.init)
            .filter { $0 != Metric.spo2.rawValue }
            .joined(separator: ",")
    }
}

File: android/app/src/main/java/com/noop/data/WhoopLiveCapabilities.kt
package com.noop.data

/**
 * Honest metrics produced by NOOP's live compatible-band path.
 *
 * Calibrated SpO2 is import-only, and the reverse-engineered motion counter is not a validated
 * pedometer. Neither may be advertised as a live capability.
 */
object WhoopLiveCapabilities {
    val base: Set<Metric> = setOf(Metric.hr, Metric.hrv, Metric.skinTemp, Metric.sleep, Metric.strainLoad)

    fun metrics(@Suppress("UNUSED_PARAMETER") model: String): Set<Metric> = base

    fun encoded(model: String): String = metrics(model).map { it.name }.sorted().joinToString(",")

    fun withoutUnvalidatedLiveMetrics(capabilities: Set<Metric>): Set<Metric> =
        capabilities - setOf(Metric.spo2, Metric.steps)

    /** Retained for the historical migration that removed only the SpO2 token. */
    fun withoutCalibratedSpo2(capabilities: Set<Metric>): Set<Metric> = capabilities - Metric.spo2

    fun stripUnvalidatedLiveTokens(encoded: String): String = encoded.split(',')
        .map { it.trim() }
        .filter { it.isNotEmpty() && it != Metric.spo2.name && it != Metric.steps.name }
        .joinToString(",")

    /** Retained for the historical migration that removed only the SpO2 token. */
    fun stripSpo2Token(encoded: String): String = encoded.split(',')
        .map { it.trim() }
        .filter { it.isNotEmpty() && it != Metric.spo2.name }
        .joinToString(",")
}
